README.md
The \strong{prevtoinc} package implements functionality to simulate point-prevalence studies (PPSs) for healtcare-associated infections (HAIs) and estimate incidence based on the PPS data. It is a companion package to the preprint [Willrich et al., From prevalence to incidence - a new approach in the hospital setting;
 doi:10.1101/554725](https://www.doi.org/10.1101/554725) ,  where methods are explained in detail.
 
 For an introduction to the use of the package have a look at the accompanying vignette 'prevtoinc_vignette'.

NEWS.md
### prevtoinc 0.12

* add new functions `calculate_CI_I_pp` and `CI_np_bs` to calculate confidence intervals for the _gren_ estimate of incidence per patient
* add new function `create_patient_history_add` to simulate a timeline of a bed
* add "naive" estimator based on empirical probability of A=1 to the set of estimators for `calculate_I` and `generate_I_fast` 
* some smaller fixes and changes

### prevtoinc 0.11

* initial CRAN-release

tests/testthat/testsestimation.R
context("Tests for estimation functions")


geom.dist.fct <- function(x) dgeom(x-1, 1/8)
geom.dist <- geom.dist.fct(1:70)/sum(geom.dist.fct(1:70))
a <- length_unbiased_mean(length_biased_dist(geom.dist))
b <- sum(1:length(geom.dist)*geom.dist)
expect_equal(a,b)


pois.dist.fct <- function(x) dpois(x-1, 7)
pois.dist <- pois.dist.fct(1:70)/sum(pois.dist.fct(1:70))
a <- length_unbiased_mean(length_biased_dist(pois.dist))
b <-sum(1:length(pois.dist)*pois.dist)
expect_equal(a,b)

expect_equal(epmf(rep(1:10,10)), rep(1/10,10))

## test conversion function _rhame
context("Testing calculate_I_rhame")

pois.dist <- create_dist_vec(pois_dist_fct, max.dist = 70)
data.pps.fast <- readRDS("refdata/datappsfast.Rds")




I.df <- calculate_I_rhame(data.pps.fast,
                          x.loi.hat = 8,
                          method = "method.identifier")
expect_known_value(I.df, 
                   "refdata/Idf.Rds",
                   update = FALSE)

I.df <- calculate_I_rhame(data.pps.fast,
                          x.loi.hat = 8,
                          x.los.hat = 6,
                          method = "method.identifier")
expect_known_value(I.df, 
                   "refdata/Idflos.Rds",
                   update = FALSE)


## test conversion function _smooth
context("Testing calculate_I_smooth (gren)")
I.df2 <- calculate_I_smooth(data.pps.fast, method = "gren")
expect_known_value(I.df2, 
                   "refdata/Idf2.Rds",
                   update= FALSE)

context("Testing calculate_I_smooth (rear)")
I.df3 <- calculate_I_smooth(data.pps.fast, method = "rear")
expect_known_value(I.df3,
                   "refdata/Idf3.Rds",
                   update = FALSE)


if (capabilities("long.double")) {
  data.pps.fast$L.loi <- NULL
  data.pps.fast$L.los <- NULL
  data.pps.fast$A.los <- NULL
  I.df4 <- calculate_I(data.pps.fast)
  expect_known_value(I.df4,
                     "refdata/Idf4.Rds",
                     update = FALSE
  )
}




R/CIfunctions.R
globalVariables(c("n",
                  "n.noso", 
                  "P.hat",
                  "x.los.hat",
                  "x.loi.hat"))



#' Calculate confidence intervals for Grenander estimator of Ipp based on asymptotics
#' 
#' Asymptotic or bootstrap approximation of confidence intervals for estimates of Ipp with gren method
#' Can use output of calculate_I_* functions as input. The asymptotic method uses 
#' the asymptotic normality of the estimator of I.pp to calculate the confidence interval and
#' the method "bs" uses a parametric bootstrap approximation based on the "naive" estimator. 
#'
#' @param data data frame which contains at least the following columns
#'  \itemize{
#'    \item{n - }{number of patients sampled} 
#'    \item{n.noso - }{number of HAIs}
#'    \item{P.hat - }{estimate of prevalence P}
#'    \item{x.loi.hat - }{estimate of x.loi}
#'    \item{x.los.hat - }{estimate of x.los}
#'    \item{I.pp.hat - }{estimate of incidence proportion per admission I.pp}
#' }
#' @param method either "asymptotic" for asymptotic confidence interval or "bs" for bootstrap-based confidence interval
#' @param alpha confidence level
#' @param n_bs number of bootstrap replications if method is "bs"
#'
#' @return tibble with columns CI.lower.Ipp and CI.upper.Ipp
#' 
#' @examples 
#' 
#' example.dist <- create_dist_vec(function(x) dpois(x-1, 7), max.dist = 70)
#' example.dist.los <- create_dist_vec(function(x) dpois(x-1, lambda = 12),
#'                                     max.dist = 70)
#' data.pps.fast <- simulate_pps_fast(n.sample=5000,
#'                                    P=0.05,
#'                                    dist.X.loi = example.dist,
#'                                    dist.X.los = example.dist.los)
#' gren_est <- calculate_I_smooth(data = data.pps.fast, method = "gren")
#' gren_est
#' calculate_CI_I_pp(gren_est, method = "asymptotic", alpha = 0.05)
#' 
#' @export
calculate_CI_I_pp <- function(data, method = "asymptotic", alpha = 0.05, n_bs = 10000) {
  if( method == "asymptotic") {
    data <- data %>% 
      dplyr::mutate(sd = sqrt(
        P.hat^2*x.los.hat^2/x.loi.hat^2*(
          1/n     *    (1-P.hat)/P.hat +
            1/n     *  (1-1/x.los.hat)*x.los.hat +
            1/n.noso * (1-1/x.loi.hat)*x.loi.hat)
      ))
      tibble::tibble(CI.lower.Ipp = data$I.pp.hat+stats::qnorm(alpha/2)*data$sd,
        CI.upper.Ipp = data$I.pp.hat+stats::qnorm(1-alpha/2)*data$sd)
  } else {
    if (method == "bs") {
      
      
      data_CI <- purrr::pmap_df(data,
                         function(n, n.noso, P.hat, x.los.hat, x.loi.hat,  ...) {
                           CI <- stats::quantile((stats::rbinom(n_bs, n, P.hat)/n)/
                                            (stats::rbinom(n_bs , n, 1/x.los.hat)/n)*
                                            (stats::rbinom(n_bs , n.noso, 1/x.loi.hat)/n.noso),
                                          c(alpha/2, 1- alpha/2))
                           
                           tibble::tibble(CI.lower.Ipp = CI[1],
                                  CI.upper.Ipp = CI[2])
                         } )
      data_CI
      
    } else stop("Method for calculating CIs unknown.")
  } 
}


#' Function to calculate confidence intervals I.pp for gren estimator with bootstrap method based on Grenander estimator
#' 
#' Implements a bootstrap procedure for estimation of confidence intervals for I.pp based on boostrapping 
#' from the length of stay/infection distributions estimated by the gren method.
#' 
#' @param data data frame which contains a column `A.loi` 
#' with lengths of nosocomial infections up to survey point ( zero if none) and
#'  a column `A.los` with length of stay up to survey point
#' @param n_bs number of bootstrap samples to use for calculations
#' @param alpha confidence level
#' 
#' @return single-row tibble with columns CI.lower.Ipp and CI.upper.Ipp
#' 
#' @examples 
#' 
#' example.dist <- create_dist_vec(function(x) dpois(x-1, 7), max.dist = 70)
#' example.dist.los <- create_dist_vec(function(x) dpois(x-1, lambda = 12),
#'                                     max.dist = 70)
#' data.pps.fast <- simulate_pps_fast(n.sample=5000,
#'                                    P=0.05,
#'                                    dist.X.loi = example.dist,
#'                                    dist.X.los = example.dist.los)
#' gren_est <- calculate_I_smooth(data = data.pps.fast, method = "gren")
#' gren_est
#' CI_np_bs(data.pps.fast, n_bs = 500)
#' @export
CI_np_bs <- function(data, n_bs = 1000, alpha = 0.05) {
  # create distribution to sample from
  noso.vec <- na.omit(data$A.loi)
  n <- length(noso.vec)
  n.noso <- length(noso.vec[noso.vec>0])
  P.hat <- sum(noso.vec > 0)/length(noso.vec)
  
  p.A.loi.bs <-
    monotone_smoother(
      noso.vec[noso.vec > 0], method = "gren")
  p.A.los.bs <-
    monotone_smoother(
      na.omit(data$A.los), method = "gren")
  
  p.A.loi.sample <-
    replicate(n_bs,
              monotone_smoother(sample.int(length(p.A.loi.bs),
                                           n.noso, replace = TRUE, p.A.loi.bs),
                                method = "gren")[1])
  
  p.A.los.sample <-
    replicate(n_bs,
              monotone_smoother(sample.int(length(p.A.los.bs),
                                           n, replace = TRUE, p.A.los.bs),
                                method = "gren")[1])
  
  P.sample <- stats::rbinom(n_bs, n, prob = P.hat)/n
  CI <- stats::quantile(P.sample*p.A.loi.sample/p.A.los.sample, c(alpha/2,1-alpha/2))
  tibble::tibble(CI.lower.Ipp = CI[1],
         CI.upper.Ipp = CI[2])

}


R/estimationfunctions.R
#' Calculate mean of length-unbiased distribution from discrete length-biased distribution 
#' starting at 1
#' @param dist vector of probabilities of length-biased distribution
#' @return mean of length-unbiased distribution 
#' 
#' @examples 
#' 
#' # geometric distribution starting in 1 and cutoff at 70 with mean at about 8.
#' geom.dist <- create_dist_vec(geom_dist_fct, max.dist = 70)
#' # calculate mean of distribution
#' sum(1:length(geom.dist)*geom.dist)
#' # create length-biased distribution in same format
#' geom.dist.lb <- length_biased_dist(geom.dist)
#' 
#' # recalculate mean of original distribution based on length-biased distribution
#' length_unbiased_mean(geom.dist.lb)
#' 
#' @export
length_unbiased_mean <- function(dist) {
  1/sum(1/1:length(dist)*dist)
}

#' Calculate length-biased distribution from discrete length-unbiased distribution
#' starting at 1
#' @param dist vector of probabilities of distribution to transform
#' @return vector of probabilities of transformed distribution
#' 
#' @examples 
#' 
#' # geometric distribution starting in 1 and cutoff at 70 with mean at about 8.
#' geom.dist <- create_dist_vec(geom_dist_fct, max.dist = 70)
#' # calculate mean
#' sum(1:length(geom.dist)*geom.dist)
#' # plot original distribution
#' plot(geom.dist)
#' geom.dist.lb <- length_biased_dist(geom.dist)
#' # plot length biased distribution
#' plot(geom.dist.lb)
#' 
#' @export
length_biased_dist <- function(dist) {
  X.mean <- sum(1:length(dist)*dist)
  dist*1:length(dist)/X.mean
}

#' Calculate empirical probability mass function
#' for a discrete positive distribution starting at 1
#' 
#' @param values used for the calculation of the empirical pmf
#' @return vector of probabilities for epmf for the range 1:length(values)
#' 
#' @examples
#' 
#' # generate random sample of independent draws from Poisson distribution
#' x <- rpois(200,4)
#' # calculate empirical probability mass function and true probability mass function
#' y.emp  <- epmf(x)
#' y.theo <- dpois(1:max(x), 4)
#' plot(y.emp)
#' points(y.theo, col = "red")
#' 
#' @export
epmf <- function(values) {
  n <- length(values)
  sapply(1:max(values), function(i) sum(values==i)/n)
}
#' Transform a distribution of times of stay to a
#' distribution of staying-time up to observation point under
#' assumption of steady state.
#'
#' @param dist.stays vector of probabilities of being at the
#' hospital for 1:length(dist.stays) days at random time of
#' observation
#' @return vector of probabilities of staying
#' 1:length(dist.point) days
#' 
#' @examples 
#' 
#' # generate vector of probabilities for truncated Poisson distribution for 
#' # distribution of times of stay X
#' dist.X <- dpois(1:70, 4)
#' plot(dist.X)
#' # transform to distribution of distribution of staying-time up to observation point under
#' # assumption of steady state
#' dist.A <- X_to_A_dist(dist.X)
#' plot(dist.A)
#' # transform back to get original distribution
#' dist.X.2 <- A_to_X_dist(dist.A)
#' plot(dist.X.2)
#' 
#' @export
X_to_A_dist <- function(dist.stays) {
  numerators <- rev(cumsum(rev(dist.stays)))
  denominator <- sum(numerators)
  return(numerators/denominator)
}

#' function to transform the distribution of stays to a fixed
#' point to the distribution of the staying times
#'
#' @param dist.point vector of probabilities of staying
#' 1:length(dist.point) days
#'
#' @return  vector of probabilities
#' of being at the hospital for 1:length(dist.point) days at
#' random time of observation
#'
#' @examples 
#' 
#' # generate vector of probabilities for truncated Poisson distribution for 
#' # distribution of times of stay X
#' dist.X <- dpois(1:70, 4)
#' plot(dist.X)
#' # transform to distribution of distribution of staying-time up to observation point under
#' # assumption of steady state
#' dist.A <- X_to_A_dist(dist.X)
#' plot(dist.A)
#' # transform back to get original distribution
#' dist.X.2 <- A_to_X_dist(dist.A)
#' plot(dist.X.2)
#' 
#' @export
A_to_X_dist <- function(dist.point) {
  numerators <- dist.point - c(dist.point[2:length(dist.point)],
                               0)
  denominator <- sum(numerators)
  return(numerators/denominator)
}

#' Calculate a monotone probability mass function estimate
#'
#' using a  rearrangement or a Grenander estimator as described in Jankoswski,
#' Wellner, 2009 <doi:10.1214/09-EJS526>
#'
#' @param values observed values of distribution
#' @param method method of estimation "rear" rearrangement or "gren" Grenander
#' @param range boundaries of the support of the distribution
#'
#' @return vector of estimated pmf (support of distribution is by default assumed to be min(values):max(values) )
#'
#' @examples
#' 
#' # generate sample from geometric distribution
#' A <- rgeom(50, 0.2)
#' # plot empirical probability mass function
#' plot(epmf(A))
#' dist.A.gren <- monotone_smoother(A, method = "gren")
#' # plot estimated probability mass function
#' points(dist.A.gren, col = "red")
#'
#' @export
monotone_smoother <- function(values, method = "rear", 
                              range = c(1, max(values))) {
  p <- sapply(1:range[2], function(i) sum(values == i))
  p <- p/length(values)
  if (method == "gren") {
    return(.find_gren(p))
  } else if (method == "rear") {
    return(.find_rear(p)) 
    } else {
    stop("monotone smoother method not known")
  }
}

.find_rear <- function(p) {
  sort(p, decreasing = TRUE)
}


# R function to apply Grenander estimator
.find_gren <- function(p) {
  -stats::isoreg(1:length(p),-p)$yf
}



R/calculateIfunctions.R
#' Estimate the incidence based on PPS data using monotone estimators for the distribution of A.
#' 
#' Estimate incidence from PPS by the method proposed in the companion paper.
#' `data` should be supplied as a data frame with at least a column named `A.loi`
#' giving lengths of infection up to date of PPS.
#' Values of zero for `A.loi` indicate absence of a HAI. Optionally,
#' the data frame can also contain a column `A.los` supplying lengths
#' of stay up to PPS to estimate x.los with the same method as well.
#' If `correct.one` is `TRUE`, the number infections on their first day
#' will be augmented to be at least as high as the number of infections
#' on their second day for the estimation of x.loi .
#'
#' @param data data frame which contains a column `A.loi` 
#' with lengths of nosocomial infections up to survey point ( zero if none) and
#' possibly a column `A.los` with length of stay up to survey point
#' @param method method to use for smoothing ("gren" ( Grenander ) or "rear" (rearrangement))
#' @param correct.one.loi use correction for underreporting of one day LOIs:
#'  "no" if none, "fill.ones" to set the one-day cases to be at least the number
#'   of two-day cases, "start.two" to only use P(A=2| A > 1) as a proxy for P(A=1)
#' @param correct.one.los use correction for underreporting of one day LOSs:
#'  "no" if none, "fill.ones" to set the one-day cases to be at least the number
#'   of two-day cases, "start.two" to only use P(A=2| A > 1) as a proxy for P(A=1)
#'
#' @return one-row data frame with following columns \itemize{
#'    \item{n - }{number of patients sampled} 
#'    \item{n.noso - }{number of HAIs}
#'    \item{P.hat - }{estimate of prevalence P}
#'    \item{I.hat - }{estimate of incidence rate I}
#'    \item{I.pp.hat - }{estimate of incidence proportion per admission I.pp}
#'    \item{x.loi.hat - }{estimate of x.loi}
#'    \item{x.los.hat - }{estimate of x.los}
#'    \item{method - }{name of the method}
#' }
#' 
#' @examples
#' 
#' # create example data for PPS
#' example.dist <- create_dist_vec(function(x) dpois(x-1, 7), max.dist = 70)
#' example.dist.los <- create_dist_vec(function(x) dpois(x-1, lambda = 12),
#'                                     max.dist = 70)
#' data.pps.fast <- simulate_pps_fast(n.sample=200,
#'                                    P=0.05,
#'                                    dist.X.loi = example.dist,
#'                                    dist.X.los = example.dist.los)
#' head(data.pps.fast)
#' 
#' # estimate of incidence
#' calculate_I_smooth(data = data.pps.fast,
#'                    method = "gren")
#' 
#' 
#' @export
calculate_I_smooth <- function(data,
                               method = "gren",
                               correct.one.loi="no",
                               correct.one.los="no") {
  # vectors of length of stay up to date of survey and length
  # of nosocomial infection up to survey
  noso.vec <- na.omit(data$A.loi)
  n <- length(noso.vec)

  P.hat <- sum(noso.vec > 0)/length(noso.vec)
  P.odds <- P.hat/(1 - P.hat)
  if (length(noso.vec[noso.vec>0]) == 0) {
    I.hat <- 0; x.loi.hat <- NA
  } else {
    if (correct.one.loi == "fill.ones") { 
      noso.vec <- c(noso.vec,
                    rep(1, max(0,sum(noso.vec==2) -
                                 sum(noso.vec==1))))
    } else if (correct.one.loi == "start.two") {
      noso.vec <- noso.vec - 1
      noso.vec[noso.vec == -1] <- 0
    }
    
    n.noso <- length(noso.vec[noso.vec>0])
    
    p.A <-
      monotone_smoother(
        noso.vec[noso.vec > 0], method = method)
    x.loi.hat <- 1/p.A[1]
    if (correct.one.loi == "start.two") x.loi.hat <- x.loi.hat + 1
    I.hat <- P.odds/x.loi.hat
  }
  
  if ("A.los" %in% colnames(data) && any(!is.na(data$A.los))) {
    los.vec <- na.omit(data$A.los)
    if (correct.one.los == "fill.ones") { 
      los.vec <- c(los.vec,
                   rep(1, max(0,sum(los.vec==2) -
                                sum(los.vec==1))))
    } else if (correct.one.los == "start.two") {
      los.vec <- los.vec - 1
      los.vec[los.vec == -1] <- 0
    }
    
    p.A.los <-
      monotone_smoother(
        los.vec[los.vec > 0],
        method = method)
    x.los.hat <- 1/p.A.los[1]
    if (correct.one.los == "start.two") x.los.hat <- x.los.hat + 1
  } else {
    x.los.hat <- NA
  }
  I.pp.hat <- I.hat*(1-P.hat)*x.los.hat
  
  return(tibble::tibble(
    n = n,
    n.noso = n.noso,
    P.hat = P.hat,
    I.hat = I.hat,
    I.pp.hat = I.pp.hat,
    x.loi.hat = x.loi.hat,
    x.los.hat = x.los.hat,
    method = method))
}


#' Function to calculate incidence from PPS data using a Rhame-Sudderth like approach 
#' with estimates for x.loi and x.los supplied.
#'
#'
#' @param data one-row data frame which contains a column A.loi (only used to calculate P.hat) 
#' with lengths of nosocomial infections up to survey (a 0 indicates no HAI present)
#' @param x.loi.hat value for estimated expected length of infection x_loi
#' @param x.los.hat value for estimated expected length of stay x_los (optional)
#' @param method a string with associated name for method
#'
#' @return one-row data frame with following columns \itemize{
#'    \item{n - }{number of patients sampled} 
#'    \item{n.noso - }{number of HAIs}
#'    \item{P.hat - }{estimate of prevalence P}
#'    \item{I.hat - }{estimate of incidence rate I}
#'    \item{I.pp.hat - }{estimate of incidence proportion per admission I.pp}
#'    \item{x.loi.hat - }{estimate of x.loi}
#'    \item{x.los.hat - }{estimate of x.los}
#'    \item{method - }{name of the method}
#' }
#' 
#' @examples
#' 
#' # create example data for PPS
#' example.dist <- create_dist_vec(function(x) dpois(x-1, 7), max.dist = 70)
#' example.dist.los <- create_dist_vec(function(x) dpois(x-1, lambda = 12),
#'                                     max.dist = 70)
#' data.pps.fast <- simulate_pps_fast(n.sample=200,
#'                                    P=0.05,
#'                                    dist.X.loi = example.dist,
#'                                    dist.X.los = example.dist.los)
#' head(data.pps.fast)
#' 
#' # estimate incidence based on Rhame-Sudderth formula
#' calculate_I_rhame(data = data.pps.fast,
#'                    x.loi.hat = 8,
#'                    x.los.hat = 13)
#' 
#' @export
calculate_I_rhame <- function(data, 
                              x.loi.hat, 
                              x.los.hat = NA, 
                              method = "rhame") {
  # vectors of length of stay up to date of survey and length
  # of nosocomial infection up to survey
  n = nrow(data)
  noso.vec <- na.omit(data$A.loi)
  n.noso <- length(noso.vec[noso.vec>0])
  
  P.hat <- sum(noso.vec > 0)/length(noso.vec)
  
  I.hat <- P.hat/(1-P.hat)/x.loi.hat
  I.pp.hat <- I.hat*(1-P.hat)*x.los.hat
  
  return(tibble::tibble(
    n = n,
    n.noso = n.noso,
    P.hat = P.hat,
    I.hat = I.hat,
    I.pp.hat = I.pp.hat,
    x.loi.hat = x.loi.hat,
    x.los.hat = x.los.hat,
    method = method)
  )
}


#' Function to calculate incidence from PPS data using a mix of two estimators 
#' 
#' A sigmoid function with parameters a and b (see below) is used to get weights
#' for a combination of the two estimator for x.loi and x.los.
#'
#'  is achieved in the following way for estimation of x.loi
#' alpha = exp(a*(n.noso-b))/(1+exp(a*(n.noso-b)))
#' x.loi.hat.mixed = alpha*x.loi.hat.1 + (1-alpha)*x.loi.hat.2
#' 
#' alpha = exp(a*(n-b))/(1+exp(a*(n-b)))
#' x.los.hat.mixed = alpha*x.los.hat.1 + (1-alpha)*x.los.hat.2
#'
#' @param I.pps.1 resulting data frame for first estimator
#' @param I.pps.2 resulting data frame for second estimator
#' @param a parameter a for the sigmoid function 
#' @param b parameter b for the sigmoid function
#' @param method name of the method
#'
#' @return one-row data frame with following columns \itemize{
#'    \item{n - }{number of patients sampled} 
#'    \item{n.noso - }{number of HAIs}
#'    \item{P.hat - }{estimate of prevalence P}
#'    \item{I.hat - }{estimate of incidence rate I}
#'    \item{I.pp.hat - }{estimate of incidence proportion per admission I.pp}
#'    \item{x.loi.hat - }{estimate of x.loi}
#'    \item{x.los.hat - }{estimate of x.los}
#'    \item{method - }{name of the method}
#' }
#' 
#' @examples 
#' 
#' # create example data for PPS
#' example.dist <- create_dist_vec(function(x) dpois(x-1, 7), max.dist = 70)
#' example.dist.los <- create_dist_vec(function(x) dpois(x-1, lambda = 12),
#'                                     max.dist = 70)
#' data.pps.fast <- simulate_pps_fast(n.sample=200,
#'                                    P=0.05,
#'                                    dist.X.loi = example.dist,
#'                                    dist.X.los = example.dist.los)
#' head(data.pps.fast)
#' 
#' # estimate of incidence
#' I.1 <- calculate_I_smooth(data = data.pps.fast,
#'                    method = "gren")
#'                    
#' # estimate incidence based on Rhame-Sudderth formula
#' I.2 <- calculate_I_rhame(data = data.pps.fast,
#'                    x.loi.hat = 8,
#'                    x.los.hat = 13)
#' 
#' # mixed estimator                                      
#' calculate_I_mixed(I.1, I.2)
#' 
#' @export
calculate_I_mixed <- function(I.pps.1, I.pps.2, a = 0.01, b = 500, 
                              method = "pps.mixed") {
  method.name <- method
  alpha.loi <- exp(a*(I.pps.1$n.noso-b))/(1+exp(a*(I.pps.1$n.noso-b)))
  alpha.los <- exp(a*(I.pps.1$n-b))/(1+exp(a*(I.pps.1$n-b)))
  I.new.mixed <- I.pps.1 %>%
    dplyr::mutate(x.los.hat = (1-alpha.los)*.data$x.los.hat+
                    alpha.los*I.pps.2$x.los.hat,
                  x.loi.hat = (1-alpha.loi)*.data$x.loi.hat+
                    alpha.loi*I.pps.2$x.loi.hat) %>%
    dplyr::mutate(
      I.hat = .data$P.hat/(1-.data$P.hat)/.data$x.loi.hat,
      I.pp.hat = .data$P.hat/.data$x.loi.hat*.data$x.los.hat) %>%
    dplyr::mutate(method = method.name)
  I.new.mixed
}